addons/stock_account/models/product.py

```diff
@@ -749,8 +749,13 @@ def _compute_average_price(self, qty_invoiced, qty_to_invoice, stock_moves, is_r
         candidates = stock_moves\
             .sudo()\
             .filtered(lambda m: is_returned == bool(m.origin_returned_move_id and sum(m.stock_valuation_layer_ids.mapped('quantity')) >= 0))\
-            .mapped('stock_valuation_layer_ids')\
-            .sorted()
+            .mapped('stock_valuation_layer_ids')
+
+        if len(candidates) > 1:
+            candidates = candidates.sorted()
+
+        if self._context.get('prefetch_candidates'):
+            candidates = candidates.with_prefetch(self._context.get('prefetch_candidates'))
```

Comment:
So this is the tricky part:
In `_accumulate_amounts`, you have all the val layers you care about `candidates`, and in `_compute_average_price`, the new layer you are getting from the `filtered` just above has no prefetch (only itself). Basically you are forwarding the prefetch_ids you've already computed in the previous context of POS, but `_compute_average_price` is a generic function unaware of the previous "precomputations". Maybe this is a symptom of a necessity of a change of signature for this compute (not for stable).

Side note that for X2many, the prefetch_ids are a special instances of a class called PrefetchX2many. 

Forwarding the prefetch ids helps in subsequent code, like `_consume_specific_qty`, for example.

Regarding generic code remarks:
- Avoid reffering to `_context`, but use `env.context`, which is the public API the ORM exposes. This gives flexibility for eventual changes of the internal representation, if in the future `_context` moves from the recordset to *somewhere else*.
- maybe the name `prefetch_candidates` is not really great? Maybe `candidates_prefetch_ids` is better namespacing. The prefix `prefetch_` would be reserved for something from the ORM itself, like `prefetch_fields` for example.